### PR TITLE
propose, nodejs compatibility

### PR DESCRIPTION
**Pull request by maboiteaspam**
Hi @rvagg 

Not sure you really want this, please let me know, i ll make a fork then.

It replaces your ```tsml``` module by ```underscore.template``` module.

By running it againt its repository, i found sometimes ```commit.summary``` was not set. 

I fixed that dirtily, i believe this way in a way it won t break any es6 things i m not aware.

```jshint``` did not complain anything here.

### CODE REVIEW

**Comment by rvagg** [comment]
`const` should be available everywhere fwiw

a couple of style things I'd like fixed up before I properly review, will comment inline on those

File: changelog-maker.js [hunk 1]
```diff
@@ -1,13 +1,14 @@
 #!/usr/bin/env node
 
-const gitcmd        = 'git log --pretty=full --since="{{sincecmd}}" --until="{{untilcmd}}"'
+var gitcmd        = 'git log --pretty=full --since="{{sincecmd}}" --until="{{untilcmd}}"'
```
Comment (rvagg) [inline changelog-maker.js]: please line up the `=` here and elsewhere you've made changes, also as I mentioned `const` should not be a problem (I've been using it for ages)

File: changelog-maker.js [hunk 2]
```diff
@@ -28,6 +28,10 @@ const spawn    = require('child_process').spawn
       }
 
 
+    _.templateSettings = {
```
Comment (rvagg) [inline changelog-maker.js]: don't indent this

File: changelog-maker.js [hunk 3]
```diff
@@ -73,7 +77,8 @@ function commitTags (list, callback) {
     sublist.forEach(function (commit) {
       function onFetch (err, issue) {
         if (err) {
-          console.error(`Error fetching issue #${commit.ghIssue}: ${err.message}`)
+          console.error(_.template('Error fetching issue #${commit.ghIssue}: ${err.message}')
+          ({commit: commit, err: err}) );
```
Comment (rvagg) [inline changelog-maker.js]: indent

File: changelog-maker.js [hunk 4]
```diff
@@ -95,7 +100,7 @@ var revertRe = /^revert\s+"?/i
   , groupRe  = /^((:?\w|\-|,|, )+):\s*/i
 
 function commitToGroup (commit) {
-  var summary = commit.summary.replace(revertRe, '')
+  var summary = (''+commit.summary).replace(revertRe, '')
```
Comment (rvagg) [inline changelog-maker.js]: whitespace around `+` please, or `String(commit.summary)` or something more elegant

File: changelog-maker.js [hunk 5]
```diff
@@ -109,18 +114,16 @@ function cleanMarkdown (txt) {
...
-    ${data.revert ? 'Revert "' : ''}
-    ${data.group ? data.group + ': ' : ''}
-    ${data.summary}
-    ${data.revert ? '"' : ''} 
-    ${data.author ? '(' + data.author + ') ' : ''}
-    ${data.pr}
-
-  `
+  var s = _.template(
+  "* [${data.sha.substr(0, 10)}] - " +
```
Comment (rvagg) [inline changelog-maker.js]: indent the internals of the call to `template()`

File: changelog-maker.js [hunk 6]
```diff
@@ -155,15 +156,16 @@ function toStringMarkdown (data) {
 function commitToOutput (commit) {
   var data       = {}
     , prUrlMatch = commit.prUrl && commit.prUrl.match(/^https?:\/\/.+\/([^\/]+\/[^\/]+)\/\w+\/\d+$/i)
+    , templateData = {ghUser: ghUser, ghProject: ghProject, commit: commit, prUrlMatch: prUrlMatch}
```
Comment (rvagg) [inline changelog-maker.js]: spaces around `{` and `}` please

File: package.json [hunk 7]
```diff
@@ -19,7 +19,7 @@
     "split2": "~0.2.1",
     "strip-ansi": "~2.0.1",
     "through2": "~0.6.3",
-    "tsml": "~1.0.0"
+    "underscore": "^1.8.3"
```
Comment (rvagg) [inline package.json]: no `^` please

File: changelog-maker.js (hunk 5 above)
Comment (rvagg) [inline changelog-maker.js]: also, replace all uses of `"` with `'` please

**Comment by maboiteaspam** [comment]
hope that works, i also fine tuned a little bit

**Comment by maboiteaspam** [comment]
Hi,

i made more significant changes in a separate branch, here
https://github.com/maboiteaspam/changelog-maker/tree/better-project-id-detection 

See a relevant commit message here for an overview of the changes
https://github.com/maboiteaspam/changelog-maker/commit/196426be10541b829112fcb388fd624edff53491

Do i merge that into this ?

File: changelog-maker.js [hunk 8]
```diff
@@ -109,18 +107,15 @@ function cleanMarkdown (txt) {
...
-    ${data.semver.length ? '(' + data.semver.join(', ').toUpperCase() + ') ' : ''}
-    ${data.revert ? 'Revert "' : ''}
-    ${data.group ? data.group + ': ' : ''}
-    ${data.summary}
-    ${data.revert ? '"' : ''} 
-    ${data.author ? '(' + data.author + ') ' : ''}
-    ${data.pr}
-
-  `
+  var s = '';
```
Comment (rvagg) [inline changelog-maker.js]: sans semicolon pls

File: changelog-maker.js [hunk 9]
```diff
@@ -131,18 +126,15 @@ function toStringSimple (data) {
...
-    ${data.semver.length ? '**(' + data.semver.join(', ').toUpperCase() + ')** ' : ''}
-    ${data.revert ? '***Revert*** "' : ''}
-    ${data.group ? '**' + data.group + '**: ' : ''}
-    ${cleanMarkdown(data.summary)}
-    ${data.revert ? '"' : ''} 
-    ${data.author ? '(' + data.author + ') ' : ''}
-    ${data.pr ? '[' + data.pr + '](' + data.prUrl + ')' : ''}
-
-  `
+  var s = '';
```
Comment (rvagg) [inline changelog-maker.js]: ;

File: changelog-maker.js [hunk 10]
```diff
@@ -155,15 +147,17 @@ function toStringMarkdown (data) {
 function commitToOutput (commit) {
   var data       = {}
     , prUrlMatch = commit.prUrl && commit.prUrl.match(/^https?:\/\/.+\/([^\/]+\/[^\/]+)\/\w+\/\d+$/i)
+    , urlHash = '#'+commit.ghIssue || commit.prUrl
```
Comment (rvagg) [inline changelog-maker.js]: line up `=`

File: changelog-maker.js [hunk 11]
```diff
@@ -155,15 +147,17 @@ function toStringMarkdown (data) {
...
-  data.shaUrl  = `https://github.com/${ghUser}/${ghProject}/commit/${commit.sha.substr(0,10)}`
+  data.shaUrl  = 'https://github.com/' + ghUser + '/' + ghProject + '/commit/' + commit.sha.substr(0,10)
   data.semver  = commit.labels && commit.labels.filter(function (l) { return l.indexOf('semver') > -1 }) || false
   data.revert  = revertRe.test(commit.summary)
   data.group   = commitToGroup(commit) || ''
-  data.summary = commit.summary.replace(revertRe, '').replace(/"$/, '').replace(groupRe, '')
+  data.summary = commit.summary && commit.summary.replace(revertRe, '').replace(/"$/, '').replace(groupRe, '')
   data.author  = (commit.author && commit.author.name) || ''
-  data.pr      = prUrlMatch && `${prUrlMatch[1] != ghUser + '/' + ghProject ? prUrlMatch[1] : ''}#${commit.ghIssue || commit.prUrl}`
+  data.pr      = prUrlMatch && ((prUrlMatch[1] != ghUrl ? prUrlMatch[1] : '')+urlHash)
```
Comment (rvagg) [inline changelog-maker.js]: spaces around the `+`

File: changelog-maker.js [hunk 12]
```diff
@@ -234,5 +228,5 @@ child.stderr.pipe(bl(function (err, _data) {
 
 child.on('close', function (code) {
   if (code)
-    throw new Error(`git command [${gitcmd}] exited with code ${code}`)
+    throw new Error('git command [' + gitcmd + '] exited with code ' + code);
```
Comment (rvagg) [inline changelog-maker.js]: ;

**Comment by rvagg** [comment]
I actually quite like the changes you made to detect the default package, I've been wondering how best to achieve that, I can live with `debug` too. Happy to merge but I'd appreciate some more work to match style, I know it's quirky but it's my quirky.

**Comment by rvagg** [comment]
i.e. happy to have those additional changes you've made, once we get the style nits right then I'll add you as a collaborator here

**Comment by maboiteaspam** [comment]
Hi,

it s fine to be demanding about the code style.

It s just i see now the `jshint` config does not seem to be inline with the comments you made.

Are you `standard` (feross/standard) user ? Or `semi-standard` ?

Please let me know i'll put that into the `npm run lint` script command.

**Comment by maboiteaspam** [comment]
I checked this source code against `standard` / `semi-standard`. None of them really fit the style in this repository. It s fine to me, i like their initiative, not their code style.

Then i checked my words and the jshint i have here would not catch semi colons. Did not try to tune its config, instead i gave a try to eslint, because i prefer its output output, with the help of https://github.com/valorkin/eslint-rules-mapper it was real quick.

I then adjusted the existing rules to pass the source code. It s bit unusual to proceed that way.

you can check the whole config, 
here https://github.com/maboiteaspam/changelog-maker/blob/eslint/.eslintrc
I have put my specific adjustments to pass the lint into a specific commit here 
https://github.com/maboiteaspam/changelog-maker/commit/83ae289ce090ae4e0a8d1e09df8f51ae6f627000

Let me know, at the end, my belief is that manual code style checking does not worth a cent. I d really prefer have one linter config to comply to.

**Comment by rvagg** [comment]
so many errors running eslint with that .eslintrc, do I need to be using a pre-release version or something to make it happy?

File: changelog-maker.js [hunk 13]
```diff
@@ -1,32 +1,48 @@
...
-    , ghProject     = argv._[1] || 'io.js'
+    , pkg      = require('./package.json')
+    , debug    = require('debug')(pkg.name)
+
+    , cwd      = process.cwd()
+    , pkgFile  = path.join(cwd, 'package.json')
+    , pkgData  = fs.existsSync(pkgFile) ? require(pkgFile) : {}
+    , pkgId    = pkgtoId(pkgData)
+
+    , ghId          = {
```
Comment (rvagg) [inline changelog-maker.js]: pull the `=` back a few to line up

File: changelog-maker.js [hunk 14]
```diff
@@ -1,32 +1,48 @@
...
+      , name: argv._[1] || pkgId.name || 'io.js'
+    }
     , authOptions   = {
           configName : 'changelog-maker'
         , scopes     : []
-      }
+    }
+
+const gitcmd      = 'git log --pretty=full --since="{{sincecmd}}" --until="{{untilcmd}}"'
+  , commitdatecmd = '$(git show -s --format=%cd `{{refcmd}}`)'
```
Comment (rvagg) [inline changelog-maker.js]: indent these by 2 more, you'll have to move the `=` out by the same too

**Comment by rvagg** [comment]
only a few minor style nits and we can land this

**Comment by maboiteaspam** [comment]
Hope it s fine.

**Comment by rvagg** [comment]
sorry to be a pain, but could you squash this all into a single commit please? there's a heap of noise here. If it's too difficult to squash with the rebased activity in the middle then perhaps submit a new PR with a single commit?

**Comment by maboiteaspam** [comment]
Agreed, i found also all those commits was not satisfying. Need to check how to do that.

**Comment by maboiteaspam** [comment]
opened #16, closing #12.

**Pull request closed without merging**